src/dataflow-types/src/client/fat_client.rs

```diff
@@ -0,0 +1,261 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! A client that maintains summaries of the involved objects.
+
+use crate::client::{
+    Client, Command, ComputeCommand, ComputeResponse, Response, StorageCommand, StorageResponse,
+};
+use expr::GlobalId;
+use repr::Timestamp;
+use timely::progress::{frontier::AntichainRef, Antichain, ChangeBatch};
+
+/// A client that maintains soft state and validates commands, in addition to forwarding them.
+pub struct FatClient<C> {
+    /// The underlying client,
+    client: C,
+    /// Sources that have been created.
+    ///
+    /// A `None` variant means that the source was dropped before it was first created.
+    source_descriptions: std::collections::BTreeMap<GlobalId, Option<crate::sources::SourceDesc>>,
+    /// Tracks `since` and `upper` frontiers for indexes and sinks.
+    compute_since_uppers: SinceUpperMap,
+    /// Tracks `since` and `upper` frontiers for sources and tables.
+    storage_since_uppers: SinceUpperMap,
+}
+
+#[async_trait::async_trait]
+impl<C: Client> Client for FatClient<C> {
+    async fn send(&mut self, cmd: Command) {
+        self.send(cmd).await
+    }
+    async fn recv(&mut self) -> Option<Response> {
+        self.recv().await
+    }
+}
+
+impl<C: Client> FatClient<C> {
+    pub async fn send(&mut self, cmd: Command) {
+        match &cmd {
+            Command::Storage(StorageCommand::CreateSources(bindings)) => {
+                // Maintain the list of bindings, and complain if one attempts to either
+                // 1. create a dropped source identifier, or
+                // 2. create an existing source identifier with a new description.
+                for (id, description) in bindings.iter() {
+                    let description = Some(description.clone());
+                    match self.source_descriptions.get(&id) {
+                        Some(None) => {
+                            panic!("Attempt to recreate dropped source id: {:?}", id);
+                        }
+                        Some(prior_description) => {
+                            if prior_description != &description {
```

Comment:
What is the case where we re-send the create source command w/ the same description?

Reply:
The main one that has come up is "coordinator reboots, and doesn't know if its previous source creation succeeded or not". It is the sort of thing that could be made into an error once STORAGE has durable state and the coordinator can instead ask it for the existing sources and do a diff, but in the meantime just complain if there is a rebinding.